File: src/javasource/Q5_MeetingRoom2.java
package javasource;

import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Queue;

public class Q5_MeetingRoom2 {

    public static void main(String[] args) {
        Q5_MeetingRoom2 a = new Q5_MeetingRoom2();

        // false 예제
        Interval intervals3 = new Interval(0, 30);
        Interval intervals2 = new Interval(5, 10);
        Interval intervals1 = new Interval(15, 20);

/*
        // true 예제
        Interval intervals1 = new Interval(2, 4);
        Interval intervals2 = new Interval(7, 10);
*/
        Interval[] intervals = {intervals1, intervals2, intervals3};
        System.out.println(a.solve(intervals));
    }

    private int solve(Interval[] intervals) {
        if(intervals == null || intervals.length == 0) return 0;

        Arrays.sort(intervals, (a, b) -> a.start - b.start);

        Queue<Interval> minHeap = new PriorityQueue<>(intervals.length, (a, b) -> a.end - b.end);
        int max = 0;
        for (int i = 0; i < intervals.length; i++) {
            while(!minHeap.isEmpty() && minHeap.peek().end >= intervals[i].start) {
                minHeap.poll();
            }
            minHeap.offer(intervals[i]);
            max = Math.max(max, minHeap.size());
        }
        return max;
    }

}

File: src/kotlinsource/Q5_MeetingRoom2.kt
package kotlinsource

import java.lang.Integer.max
import java.util.*

fun main(args:Array<String>) {
/*
    val int1:Interval = Interval(0, 30)
    val int2:Interval = Interval(15, 20)
    val int3:Interval = Interval(5, 10)

 */
/*
    val int1:Interval = Interval(1, 4)
    val int2:Interval = Interval(4, 5)
    val int3:Interval = Interval(4, 6)
*/


    val int1:Interval = Interval(2, 4)
    val int2:Interval = Interval(7, 10)


    val list = arrayOf(int1, int2)

    print(solve(list))

}

fun solve(intervals: Array<Interval>): Int {
    if (intervals == null || intervals.isEmpty()) return 0

    val intervals = intervals.sortedBy { it.start }

    val minHeap:Queue<Interval> = PriorityQueue<Interval>(intervals.size, compareBy { it.end })
    var max: Int = 0

    for (interval in intervals) {
        while(!minHeap.isEmpty() && minHeap.peek().end <= interval.start) {
            minHeap.poll();
        }
        minHeap.offer(interval)
        max = max(max, minHeap.size)
    }
    return max
}